#!/usr/bin/env python3
import grpc
import sys
import os
import struct

# –î–æ–±–∞–≤–ª—è–µ–º src –≤ PYTHONPATH
current_dir = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, current_dir)

# –ò—Å–ø—Ä–∞–≤–ª—è–µ–º –∏–º–ø–æ—Ä—Ç, –æ–±—Ö–æ–¥—è –ø—Ä–æ–±–ª–µ–º–Ω—ã–π src/__init__.py
try:
    # –ü—Ä—è–º–æ–π –∏–º–ø–æ—Ä—Ç –∏–∑ grpc –ø–∞–ø–∫–∏, –æ–±—Ö–æ–¥—è src –ø–∞–∫–µ—Ç
    grpc_path = os.path.join(current_dir, 'src', 'grpc')
    if grpc_path not in sys.path:
        sys.path.insert(0, grpc_path)
    
    import dbc_pb2
    import dbc_pb2_grpc
    print("‚úÖ Protobuf imports successful")
    
except ImportError as e:
    print(f"‚ùå Import error: {e}")
    print("üîß Checking if protobuf files exist...")
    
    grpc_dir = os.path.join(current_dir, 'src', 'grpc')
    pb2_file = os.path.join(grpc_dir, 'dbc_pb2.py')
    pb2_grpc_file = os.path.join(grpc_dir, 'dbc_pb2_grpc.py')
    
    if not os.path.exists(pb2_file):
        print(f"‚ùå Missing: {pb2_file}")
        print("üí° Run: python3 setup.py")
        sys.exit(1)
    
    if not os.path.exists(pb2_grpc_file):
        print(f"‚ùå Missing: {pb2_grpc_file}")
        print("üí° Run: python3 setup.py")
        sys.exit(1)
    
    print("Files exist, trying alternative import...")
    
    # –ê–ª—å—Ç–µ—Ä–Ω–∞—Ç–∏–≤–Ω—ã–π —Å–ø–æ—Å–æ–± –∏–º–ø–æ—Ä—Ç–∞
    try:
        spec_pb2 = importlib.util.spec_from_file_location("dbc_pb2", pb2_file)
        dbc_pb2 = importlib.util.module_from_spec(spec_pb2)
        spec_pb2.loader.exec_module(dbc_pb2)
        
        spec_pb2_grpc = importlib.util.spec_from_file_location("dbc_pb2_grpc", pb2_grpc_file)
        dbc_pb2_grpc = importlib.util.module_from_spec(spec_pb2_grpc)
        spec_pb2_grpc.loader.exec_module(dbc_pb2_grpc)
        
        print("‚úÖ Alternative import successful")
    except Exception as e2:
        print(f"‚ùå Alternative import failed: {e2}")
        sys.exit(1)

def create_test_frame(dev_addr: int, msg_id: int, can_data: bytes) -> str:
    """–°–æ–∑–¥–∞–Ω–∏–µ —Ç–µ—Å—Ç–æ–≤–æ–≥–æ 12-–±–∞–π—Ç–æ–≤–æ–≥–æ —Ñ—Ä–µ–π–º–∞"""
    # –§–æ—Ä–º–∏—Ä—É–µ–º COMM_ADDR
    comm_addr = (dev_addr & 0x1F) | ((msg_id & 0x3FF) << 5)
    comm_addr_bytes = struct.pack('<H', comm_addr)
    
    # –î–æ–ø–æ–ª–Ω—è–µ–º CAN –¥–∞–Ω–Ω—ã–µ –¥–æ 8 –±–∞–π—Ç
    padded_data = can_data + b'\x00' * (8 - len(can_data))
    
    # –í—ã—á–∏—Å–ª—è–µ–º CRC16 (–ø–æ–∫–∞ –∑–∞–≥–ª—É—à–∫–∞)
    payload = comm_addr_bytes + padded_data
    crc16 = 0x1234  # –ó–∞–≥–ª—É—à–∫–∞ –¥–ª—è —Ç–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏—è
    crc16_bytes = struct.pack('<H', crc16)
    
    frame = payload + crc16_bytes
    return frame.hex().upper()

def test_dbc_service():
    """–¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ gRPC DBC Service"""
    
    try:
        with grpc.insecure_channel('localhost:50051') as channel:
            stub = dbc_pb2_grpc.DBCParserServiceStub(channel)
            
            print("–¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ Python DBC Service —á–µ—Ä–µ–∑ gRPC...")
            
            # Health check
            try:
                health = stub.HealthCheck(dbc_pb2.Empty())
                print(f"‚úÖ Health: {health.status}, DBC loaded: {health.dbc_loaded}, Uptime: {health.uptime_seconds}s")
            except grpc.RpcError as e:
                print(f"‚ùå Service unavailable: {e.details()}")
                print("üí° Make sure the server is running: python3 main.py")
                return
            
            # DBC info
            try:
                info = stub.GetDBCInfo(dbc_pb2.Empty())
                print(f"‚úÖ DBC Info: {info.dbc_file}, Messages: {info.total_messages}")
                print(f"   Loaded at: {info.loaded_at}")
                
                # –ü–∞—Ä—Å–∏–º –¥–æ—Å—Ç—É–ø–Ω—ã–µ —Å–æ–æ–±—â–µ–Ω–∏—è
                import json
                messages = json.loads(info.messages_json)
                print("   Available messages:")
                for msg_id, msg_name in messages.items():
                    print(f"     ID {msg_id}: {msg_name}")
            except grpc.RpcError as e:
                print(f"‚ùå Failed to get DBC info: {e.details()}")
                return
            
            print("\n" + "="*50)
            
            test_cases = [
                {
                    "name": "evse_session",
                    "id": 17,
                    "data": struct.pack('<HHH', 42, 125, 15500)[:6],  # –û–±—Ä–µ–∑–∞–µ–º –¥–æ 6 –±–∞–π—Ç —Å–æ–≥–ª–∞—Å–Ω–æ DBC
                    "description": "–î–∞–Ω–Ω—ã–µ —Å–µ—Å—Å–∏–∏ –∑–∞—Ä—è–¥–∫–∏"
                },
                {
                    "name": "evse_state", 
                    "id": 16,
                    "data": struct.pack('<BBBBBBBB', 0x03, 0x12, 0x00, 0x45, 0x23, 0x01, 0x00, 0x00),  # 8 –±–∞–π—Ç
                    "description": "–°–æ—Å—Ç–æ—è–Ω–∏–µ —É—Å—Ç—Ä–æ–π—Å—Ç–≤–∞"
                },
                {
                    "name": "evse_data1",
                    "id": 19,
                    "data": struct.pack('<BBBBBBBB', 16, 15, 16, 0, 0, 0, 156, 20),  # 8 –±–∞–π—Ç (power —É–ø–∞–∫–æ–≤—ã–≤–∞–µ–º –≤ 2 –±–∞–π—Ç–∞)
                    "description": "–î–∞–Ω–Ω—ã–µ —Ç–æ–∫–æ–≤ –∏ –º–æ—â–Ω–æ—Å—Ç–∏"
                },
                {
                    "name": "evse_data2",
                    "id": 20,
                    "data": struct.pack('<HHHbb', 230, 235, 232, 45, 38),  # 8 –±–∞–π—Ç (3H + 2b = 8)
                    "description": "–î–∞–Ω–Ω—ã–µ –Ω–∞–ø—Ä—è–∂–µ–Ω–∏–π –∏ —Ç–µ–º–ø–µ—Ä–∞—Ç—É—Ä"
                },
                {
                    "name": "remote_cmd",
                    "id": 34,
                    "data": struct.pack('<BBHH', 0x01, 0x11, 120, 50000)[:6],  # –û–±—Ä–µ–∑–∞–µ–º –¥–æ 6 –±–∞–π—Ç —Å–æ–≥–ª–∞—Å–Ω–æ DBC
                    "description": "–ö–æ–º–∞–Ω–¥—ã —É–ø—Ä–∞–≤–ª–µ–Ω–∏—è"
                }
            ]
            
            for test_case in test_cases:
                print(f"\nüß™ –¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ {test_case['name']} (ID: {test_case['id']})")
                print(f"   {test_case['description']}")
                
                # –°–æ–∑–¥–∞–µ–º —Ç–µ—Å—Ç–æ–≤—ã–π —Ñ—Ä–µ–π–º
                frame_data = create_test_frame(1, test_case['id'], test_case['data'])
                print(f"   Frame data: {frame_data}")
                
                # –ü–∞—Ä—Å–∏–º —Ñ—Ä–µ–π–º
                request = dbc_pb2.ParseFrameRequest(
                    frame_data=frame_data,
                    timestamp="2025-01-15T10:30:45Z"
                )
                
                try:
                    response = stub.ParseFrame(request)
                    print(f"   ‚úÖ Device Address: {response.device_address}")
                    print(f"   ‚úÖ Message: {response.message_name} (ID: {response.message_id})")
                    print(f"   ‚úÖ CRC Valid: {response.crc_valid}")
                    print(f"   ‚úÖ Parsed: {response.parsed}")
                    
                    if response.signals_json and response.signals_json != "{}":
                        import json
                        signals = json.loads(response.signals_json)
                        print(f"   ‚úÖ Signals:")
                        for signal_name, value in signals.items():
                            print(f"      {signal_name}: {value}")
                    
                    if response.error:
                        print(f"   ‚ö†Ô∏è  Error: {response.error}")
                        
                except grpc.RpcError as e:
                    print(f"   ‚ùå Parse error: {e.details()}")
            
            print("\n" + "="*50)
            
            # –¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ batch –æ–±—Ä–∞–±–æ—Ç–∫–∏
            print("\nüß™ –¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ batch –æ–±—Ä–∞–±–æ—Ç–∫–∏...")
            batch_frames = []
            for i, test_case in enumerate(test_cases[:3]):  # –ë–µ—Ä–µ–º –ø–µ—Ä–≤—ã–µ 3
                frame_data = create_test_frame(i+1, test_case['id'], test_case['data'])
                batch_frames.append(dbc_pb2.ParseFrameRequest(
                    frame_data=frame_data,
                    timestamp=f"2025-01-15T10:3{i}:45Z"
                ))
            
            try:
                batch_request = dbc_pb2.ParseBatchRequest(frames=batch_frames)
                responses = list(stub.ParseBatch(batch_request))
                print(f"   ‚úÖ Batch processed: {len(responses)} frames")
                
                for i, response in enumerate(responses):
                    print(f"   Frame {i+1}: {response.message_name}, Parsed: {response.parsed}")
                    
            except grpc.RpcError as e:
                print(f"   ‚ùå Batch error: {e.details()}")
            
            print("\n" + "="*50)
            
            # –¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ –Ω–µ–≤–∞–ª–∏–¥–Ω—ã—Ö –¥–∞–Ω–Ω—ã—Ö
            print("\nüß™ –¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ –æ–±—Ä–∞–±–æ—Ç–∫–∏ –æ—à–∏–±–æ–∫...")
            
            error_tests = [
                {
                    "name": "–ù–µ–ø—Ä–∞–≤–∏–ª—å–Ω–∞—è –¥–ª–∏–Ω–∞",
                    "frame": "DEADBEEF",  # –°–ª–∏—à–∫–æ–º –∫–æ—Ä–æ—Ç–∫–∏–π
                },
                {
                    "name": "–ù–µ–∏–∑–≤–µ—Å—Ç–Ω—ã–π ID —Å–æ–æ–±—â–µ–Ω–∏—è", 
                    "frame": create_test_frame(1, 999, b'\x00\x01\x02\x03'),
                },
                {
                    "name": "–ù–µ–≤–∞–ª–∏–¥–Ω—ã–π hex",
                    "frame": "INVALID_HEX_DATA_HERE!",
                }
            ]
            
            for error_test in error_tests:
                print(f"\n   –¢–µ—Å—Ç: {error_test['name']}")
                try:
                    request = dbc_pb2.ParseFrameRequest(frame_data=error_test['frame'])
                    response = stub.ParseFrame(request)
                    print(f"   ‚úÖ Error handled: {response.error}")
                    print(f"   ‚úÖ Parsed: {response.parsed}")
                except grpc.RpcError as e:
                    print(f"   ‚úÖ gRPC error handled: {e.details()}")
    
    except Exception as e:
        print(f"‚ùå Unexpected error: {e}")
        import traceback
        traceback.print_exc()

def test_with_real_crc():
    """–¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ —Å –ø—Ä–∞–≤–∏–ª—å–Ω—ã–º CRC16"""
    print("\nüîß –¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ —Å –ø—Ä–∞–≤–∏–ª—å–Ω—ã–º CRC16...")
    
    try:
        # –ò–º–ø–æ—Ä—Ç–∏—Ä—É–µ–º CRC —Ñ—É–Ω–∫—Ü–∏—é
        crc_path = os.path.join(current_dir, 'src', 'dbc')
        if crc_path not in sys.path:
            sys.path.insert(0, crc_path)
        
        from crc import crc16_arc
        
        def create_valid_frame(dev_addr: int, msg_id: int, can_data: bytes) -> str:
            """–°–æ–∑–¥–∞–Ω–∏–µ —Ñ—Ä–µ–π–º–∞ —Å –ø—Ä–∞–≤–∏–ª—å–Ω—ã–º CRC16"""
            comm_addr = (dev_addr & 0x1F) | ((msg_id & 0x3FF) << 5)
            comm_addr_bytes = struct.pack('<H', comm_addr)
            
            padded_data = can_data + b'\x00' * (8 - len(can_data))
            payload = comm_addr_bytes + padded_data
            
            # –í—ã—á–∏—Å–ª—è–µ–º –ø—Ä–∞–≤–∏–ª—å–Ω—ã–π CRC16
            crc16 = crc16_arc(payload)
            crc16_bytes = struct.pack('<H', crc16)
            
            frame = payload + crc16_bytes
            return frame.hex().upper()
        
        # –¢–µ—Å—Ç —Å –≤–∞–ª–∏–¥–Ω—ã–º CRC
        valid_frame = create_valid_frame(1, 17, struct.pack('<HHH', 42, 125, 15500))
        print(f"Frame with valid CRC: {valid_frame}")
        
        with grpc.insecure_channel('localhost:50051') as channel:
            stub = dbc_pb2_grpc.DBCParserServiceStub(channel)
            
            request = dbc_pb2.ParseFrameRequest(frame_data=valid_frame)
            response = stub.ParseFrame(request)
            
            print(f"CRC Valid: {response.crc_valid}")
            print(f"Parsed: {response.parsed}")
            if response.error:
                print(f"Error: {response.error}")
    
    except ImportError as e:
        print(f"‚ùå Cannot import CRC module: {e}")
        print("üí° CRC validation test skipped")
    except grpc.RpcError as e:
        print(f"‚ùå gRPC error in CRC test: {e.details()}")
    except Exception as e:
        print(f"‚ùå Unexpected error in CRC test: {e}")

if __name__ == "__main__":
    print("üöÄ –ó–∞–ø—É—Å–∫ —Ç–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏—è DBC Service...")
    
    # –ü—Ä–æ–≤–µ—Ä—è–µ–º, —á—Ç–æ –ø—Ä–æ—Çob—É—Ñ –º–æ–¥—É–ª–∏ –∑–∞–≥—Ä—É–∂–µ–Ω—ã
    if 'dbc_pb2' not in globals() or 'dbc_pb2_grpc' not in globals():
        print("‚ùå Protobuf modules not loaded properly")
        sys.exit(1)
    
    test_dbc_service()
    test_with_real_crc()
    print("\nüéâ –¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ –∑–∞–≤–µ—Ä—à–µ–Ω–æ!")